### PR TITLE
Native Bindings

### PR DESCRIPTION
**Pull request by StephaneDelcroix**
### Description of Change ###

Allow binding to Native (non-BP) Views.

Each platform now expose `SetBinding(string propertyName, BindingBase binding)` as an extension method on the respective native views (`UIView`, `Android.View`, `FrameworkElement`).

To get 2-Way Bindings to work, `SetBinding` can take an additional `updateSourceEventName` argument. On iOS, if the Property of the `UIView` is Observable, the `updateSourceEventName` is not mandatory. On [U]WP, something similar is done in order to try to get 2-Way bindings to behave even without the updateSourceEventName.

In addition to this, the platforms also expose `SetValue(BindableProperty property, ...)` and `SetBinding(BindableProperty property, ...)`. Those will have no effect unless the view happen to be directly wrapped in a Xamarin.Forms Layout. In that case, those (attached) BindableProperties will be transferred to the wrapper, allowing this kind of code to be valid.

```
var btn = new UIKit.Button();
btn.SetValue(Grid.RowProperty, 2); 

var grid = new Xamarin.Forms.Grid ();
grid.Children.Add(btn.ToView());
```

The `SetBindingContext` is not meant to be invoked directly, but it's used by the platform.

The value of this will make more sense when the Xaml support is added.

### Bugs Fixed ###

None

### API Changes ###

Added:
**iOS**:
 - `public static void SetBinding(this UIView view, string propertyName, BindingBase binding, string updateSourceEventName = null);`
 - `public static void SetBinding(this UIView self, BindableProperty targetProperty, BindingBase binding);`
 - `public static void SetValue(this UIView target, BindableProperty targetProperty, object value);`
 - `public static void SetBindingContext(this UIView target, object bindingContext, Func<UIView, IEnumerable<UIView>> getChildren = null);`

**Android**
 - `public static void SetBinding(this global::Android.Views.View view, string propertyName, BindingBase binding, string updateSourceEventName = null);`
 - `public static void SetBinding(this global::Android.Views.View view, BindableProperty targetProperty, BindingBase binding);`
 - `public static void SetValue(this global::Android.Views.View target, BindableProperty targetProperty, object value);`
 - `public static void SetBindingContext(this global::Android.Views.View target, object bindingContext, Func<global::Android.Views.View, IEnumerable<global::Android.Views.View>> getChildren = null);`

**[U]WP**
 - `public static void SetBinding(this FrameworkElement view, string propertyName, BindingBase binding, string updateSourceEventName = null);`
 - `public static void SetBinding(this FrameworkElement view, BindableProperty targetProperty, BindingBase binding);`
 - `public static void SetValue(this FrameworkElement target, BindableProperty targetProperty, object value);`
 - `public static void SetBindingContext(this FrameworkElement target, object bindingContext, Func<FrameworkElement, IEnumerable<FrameworkElement>> getChildren = null);`
 
### PR Checklist ###

- [x] Has tests (if omitted, state reason in description)
- [x] Rebased on top of master at time of PR
- [x] Changes adhere to coding standard
- [x] Consolidate commits as makes sense

### CODE REVIEW

File: Xamarin.Forms.Core/BindableObjectProxy.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+		public WeakReference<TNativeView> TargetReference { get; set; }
+		public IList<KeyValuePair<BindableProperty, BindingBase>> BindingsBackpack { get; } = new List<KeyValuePair<BindableProperty, BindingBase>>();
+		public IList<KeyValuePair<BindableProperty, object>> ValuesBackpack { get; } = new List<KeyValuePair<BindableProperty, object>>();
+
+		public BindableObjectProxy(TNativeView target)
+		{
+			TargetReference = new WeakReference<TNativeView>(target);
+		}
+
+		public void TransferAttachedPropertiesTo(View wrapper)
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/BindableObjectProxy.cs]: This needs to be replaced by a static call

File: Xamarin.Forms.Core/NativeBindingHelpers.cs [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+namespace Xamarin.Forms
+{
+	static class NativeBindingHelpers
+	{
+		public static void SetBinding<TNativeView>(TNativeView target, string targetProperty, BindingBase bindingBase, string updateSourceEventName) where TNativeView : class
+		{
+			if (string.IsNullOrEmpty(updateSourceEventName))
+				throw new ArgumentNullException(nameof(updateSourceEventName));
+
+			var eventWrapper = new EventWrapper(target, targetProperty, updateSourceEventName);
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/NativeBindingHelpers.cs]: Do we have tests for when we try to connect to a non-existing event ? Or to a non-existing property ? It should throw (a meaningful exception) in both cases.

Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/NativeBindingHelpers.cs]: Also, do we have tests describing how the value are transferred on initial connection (on SetBinding(), native value should be applied on OneWayToSource, e.g.)

Comment (rmarinho) [inline Xamarin.Forms.Core/NativeBindingHelpers.cs]: https://github.com/xamarin/Xamarin.Forms/pull/278/files#diff-0faacf4d695182912f1ba727d3b6a67bR362 this tests for the non existing event , need test for te property

Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/NativeBindingHelpers.cs]: That's not what I meant by non-existing event. It's trying to connect to FooFired when that event doesn't exists on the target

File: Xamarin.Forms.Core/BindableObjectProxy.cs (hunk 1 above)
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/BindableObjectProxy.cs]: FIXED

File: Xamarin.Forms.Core/NativeBindingHelpers.cs (hunk 2 above)
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/NativeBindingHelpers.cs]: FIXED

File: Xamarin.Forms.Controls/ControlGalleryPages/NativeBindingGalleryPage.cs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+			label.SetBinding(Label.TextProperty, "FormsLabel");
+
+			Layout.Children.Add(buttonNav);
+
+			Layout.Children.Add(label);
+
+			Layout.Children.Add(boxView);
+			Layout.Children.Add(button);
+
+			BindingContext = ViewModel = vm; ;
```
Comment (samhouts) [inline Xamarin.Forms.Controls/ControlGalleryPages/NativeBindingGalleryPage.cs]: double semi colon

File: Xamarin.Forms.Core.UnitTests/NativeBindingTests.cs [hunk 4]
```diff
@@ -0,0 +1,487 @@
...
+
+		public event PropertyChangedEventHandler PropertyChanged;
+	}
+
+	class MockVMForNativeBinding : INotifyPropertyChanged
+	{
+		string fFoo;
+		public string FFoo {
+			get { return fFoo; }
+			set {
```
Comment (samhouts) [inline Xamarin.Forms.Core.UnitTests/NativeBindingTests.cs]: Okay, but the curly brace placement should at least be the same within the same check-in.

File: Xamarin.Forms.Core/NativeBindingHelpers.cs [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+							SetNativeValue(nativeView, targetProperty, newValue);
+				}
+			);
+		}
+
+		static void SetNativeValue<TNativeView>(TNativeView target, string targetProperty, object newValue) where TNativeView : class
+		{
+			var mi = target.GetType().GetProperty(targetProperty)?.SetMethod;
+			if (mi == null)
+				throw new InvalidOperationException(Format("Native Binding on {0}.{1} failed due to missing or innacessible property", target.GetType(), targetProperty));
```
Comment (samhouts) [inline Xamarin.Forms.Core/NativeBindingHelpers.cs]: inaccessible

File: Xamarin.Forms.Core/NativeBindingHelpers.cs [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+				return;
+			var children = getChild(target);
+			if (children == null)
+				return;
+			foreach (var child in children)
+				if (child != null)
+					SetBindingContext(child, bindingContext, getChild);
+		}
+
+		internal static void TransferbindablePropertiesToWrapper<TNativeView, TNativeWrapper>(TNativeView nativeView, TNativeWrapper wrapper)
```
Comment (samhouts) [inline Xamarin.Forms.Core/NativeBindingHelpers.cs]: uppercase B on bindable

File: Xamarin.Forms.Platform.Android/Extensions/NativeBindingExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+		{
+			NativeBindingHelpers.SetValue(target, targetProperty, value);
+		}
+
+		public static void SetBindingContext(this global::Android.Views.View target, object bindingContext, Func<global::Android.Views.View, IEnumerable<global::Android.Views.View>> getChildren = null)
+		{
+			NativeBindingHelpers.SetBindingContext(target, bindingContext, getChildren);
+		}
+
+		internal static void TransferbindablePropertiesToWrapper(this global::Android.Views.View target, View wrapper)
```
Comment (samhouts) [inline Xamarin.Forms.Platform.Android/Extensions/NativeBindingExtensions.cs]: uppercase B on bindable

**Comment by samhouts** [comment]
need to update docs

**Comment by rmarinho** [comment]
@jassmith done please review

**Comment by StephaneDelcroix** [comment]
rebased on master as of today

**Comment by StephaneDelcroix** [comment]
rebased as of today

File: Xamarin.Forms.ControlGallery.Android/ColorPicker.cs [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+using Android.Content;
+using Android.Views;
+using Android.Widget;
+using Xamarin.Forms.Platform.Android;
+using Droid = Android;
+
+namespace Xamarin.Forms.ControlGallery.Android
+{
+	public class ColorPickerView : ViewGroup
+	//, INotifyPropertyChanged
```
Comment (hartez) [inline Xamarin.Forms.ControlGallery.Android/ColorPicker.cs]: Is removing INotifyPropertyChanged intended?

Comment (rmarinho) [inline Xamarin.Forms.ControlGallery.Android/ColorPicker.cs]: Yes, just for testing

File: Xamarin.Forms.Core/NativeBindingHelpers.cs [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using Xamarin.Forms.Internals;
+
+using static System.String;
+
+namespace Xamarin.Forms
+{
+	static class NativeBindingHelpers
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/NativeBindingHelpers.cs]: This is the meaty part to review. It's actually the only important part to review as well.

File: Xamarin.Forms.Platform.Android/ColorExtensions.cs [hunk 10]
```diff
@@ -36,5 +36,10 @@ public static ColorStateList ToAndroidPreserveDisabled(this Color color, ColorSt
 			int disabled = defaults.GetColorForState(States[1], color.ToAndroid());
 			return new ColorStateList(States, new[] { color.ToAndroid().ToArgb(), disabled });
 		}
+
+		public static Color ToColor(this AColor color)
+		{
+			return Color.FromUint((uint)color.ToArgb());
+		}
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Platform.Android/ColorExtensions.cs]: @rmarinho can we drop this ?

**Pull request merged**